### PR TITLE
WIP: IDL introspection: Debug pages

### PR DESCRIPTION
**Pull request by bombela**
IDL introspection is added to debug pages. Nothing fancy, just cute HTML.

### CODE REVIEW

**Comment by bombela** [comment]
And what it looks like:

image

image

image

File: debug.go [hunk 1]
```diff
@@ -56,169 +54,50 @@ func removeDispatcherFromDebugPages(disp *Dispatcher) {
...
-		data.Dispatchers = append(data.Dispatchers, disp.Introspect())
-	}
+func (introspectionProvider) Dispatchers() []introspection.DispatcherStatus {
+	dispatchersLock.RLock()
+	defer dispatchersLock.RUnlock()
 
-	if err := pageTmpl.ExecuteTemplate(w, "Page", data); err != nil {
-		log.Printf("yarpc/debug: Failed executing template: %v", err)
+	r := make([]introspection.DispatcherStatus, len(dispatchers))
+	for idx, d := range dispatchers {
```
Comment (peter-edge) [inline debug.go]: nit: just do `i` :)

File: debug.go [hunk 2]
```diff
@@ -56,169 +54,50 @@ func removeDispatcherFromDebugPages(disp *Dispatcher) {
...
+func (introspectionProvider) PackageVersions() []introspection.PackageVersion {
+	return PackageVersions
+}
+
+func introspectDispatchers() []introspection.DispatcherStatus {
+	dispatchersLock.RLock()
+	defer dispatchersLock.RUnlock()
+
+	r := make([]introspection.DispatcherStatus, len(dispatchers))
+	for idx, d := range dispatchers {
```
Comment (peter-edge) [inline debug.go]: `i`

File: debug.go [hunk 3]
```diff
@@ -56,169 +54,50 @@ func removeDispatcherFromDebugPages(disp *Dispatcher) {
...
+		r[idx] = d.Introspect()
+	}
+	return r
+}
+
+func init() {
+	debugPages := debug.NewPages(introspectionProvider{})
+	debugPages.RegisterOn(http.DefaultServeMux)
```
Comment (peter-edge) [inline debug.go]: nit: could just one-line this
nit: could just make `var introspectionProvderInstance = introspectionProvider{}`

File: internal/debug/debug.go [hunk 4]
```diff
@@ -0,0 +1,177 @@
...
+
+// Pages holds a bag of yarpc debug pages. Binding a set of pages to an
+// introspection provider.
+type Pages struct {
+	insp  IntrospectionProvider
+	pages map[string]*page
+}
+
+func (dp *Pages) registerPage(page *page) {
+	funcmap := map[string]interface{}{
```
Comment (peter-edge) [inline internal/debug/debug.go]: I would make this a global var

File: internal/debug/debug.go [hunk 5]
```diff
@@ -0,0 +1,177 @@
...
+		"pathBase":    path.Base,
+		"wrapIDLtree": wrapIDLTree,
+	}
+
+	// We do not .Clone() the base template, but reparse it every time. Because
+	// of a race condition/memory leak when template.Clone() is used in
+	// conjunction with template blocks on Go<=1.8.
+	base := template.Must(template.New("base").Funcs(funcmap).Parse(baseHTML))
```
Comment (peter-edge) [inline internal/debug/debug.go]: I really don't like using `Must` functions outside of global variable initialization - can we return error somehow?

File: internal/debug/debug_idl.go [hunk 6]
```diff
@@ -0,0 +1,185 @@
...
+	sort.Sort(t.Modules)
+	return idlTreeHelper{dname, i + 1, t}
+}
+
+type dispatcherWithIDLModule struct {
+	Dispatcher introspection.DispatcherStatus
+	IDLModule  introspection.IDLModule
+}
+
+const idlPagePath = "/debug/yarpc/idl/"
```
Comment (peter-edge) [inline internal/debug/debug_idl.go]: put at top

File: internal/debug/debug_idl.go [hunk 7]
```diff
@@ -0,0 +1,185 @@
...
+
+var idlPage = page{
```
Comment (peter-edge) [inline internal/debug/debug_idl.go]: put at top

File: internal/debug/debug_idl.go [hunk 8]
```diff
@@ -0,0 +1,185 @@
...
+	path: idlPagePath,
+	handler: func(w http.ResponseWriter, req *http.Request, insp IntrospectionProvider) interface{} {
```
Comment (peter-edge) [inline internal/debug/debug_idl.go]: can you make this a separate function? might be nice for testing too

File: internal/debug/debug_idl.go [hunk 9]
```diff
@@ -0,0 +1,185 @@
...
+				return nil
+			}
+			selectDispatcher = parts[0]
+			selectIDL = parts[1]
+		}
+
+		dispatchers := insp.DispatchersByName(selectDispatcher)
+
+		if len(dispatchers) == 0 {
+			w.WriteHeader(404)
```
Comment (peter-edge) [inline internal/debug/debug_idl.go]: `http.StatusNotFound`

File: internal/debug/debug_idl.go [hunk 10]
```diff
@@ -0,0 +1,185 @@
...
+		}
+
+		// Here we are hunting for the idl file path to return. Dispatchers can have
+		// duplicated names. Because
+		parts := strings.SplitN(path, "/", 2)
+		var selectDispatcher string
+		var selectIDL string
+		if path != "" {
+			if len(parts) != 2 {
+				w.WriteHeader(400)
```
Comment (peter-edge) [inline internal/debug/debug_idl.go]: `http.StatusBadRequest`

File: internal/debug/debug_idl.go [hunk 11]
```diff
@@ -0,0 +1,185 @@
...
+			if m, ok := d.Procedures.IDLModuleByFilePath(selectIDL); ok {
+				idls = append(idls, dispatcherWithIDLModule{
+					Dispatcher: d,
+					IDLModule:  *m,
+				})
+			}
+		}
+
+		if len(idls) == 0 {
+			w.WriteHeader(404)
```
Comment (peter-edge) [inline internal/debug/debug_idl.go]: `http.StatusNotFound`

File: internal/debug/debug_idl.go [hunk 12]
```diff
@@ -0,0 +1,185 @@
...
+		for _, d := range idls {
+			if len(idls) > 1 {
+				fmt.Fprintf(w, "Dispatcher %q (%q):\n",
+					d.Dispatcher.Name, d.Dispatcher.ID)
+			}
+			fmt.Fprintf(w, d.IDLModule.RawContent)
+		}
+		return nil
+	},
+	html: `
```
Comment (peter-edge) [inline internal/debug/debug_idl.go]: const?

File: internal/debug/debug_root.go [hunk 13]
```diff
@@ -0,0 +1,132 @@
...
+
+package debug
+
+import (
+	"net/http"
+
+	"go.uber.org/yarpc/internal/introspection"
+)
+
+var rootPage = page{
```
Comment (peter-edge) [inline internal/debug/debug_root.go]: Same comments as other variable

**Comment by blampe** [comment]
Please add tests!

File: internal/debug/debug.go (hunk 5 above)
Comment (bombela) [inline internal/debug/debug.go]: it must never ever fail. a failure here is a programming error, and will be caught when running tests.

I am not using globals mostly for code organisation and working around the race condition/leak bug.

File: internal/debug/debug_idl.go (hunk 8 above)
Comment (bombela) [inline internal/debug/debug_idl.go]: you can test using idlPage.handler directly. I was trying to have some sort of name-spacing by using a struct like that. Let me know if you think of a better style to organize all the pages.

File: internal/debug/debug_idl.go (hunk 12 above)
Comment (bombela) [inline internal/debug/debug_idl.go]: can you expand what you mean?

**Pull request closed without merging**